### PR TITLE
Updated regex patterns to capture Spark 2.0 / Spark 2.1 metrics

### PR DESCRIPTION
**Pull request by pablete**
Updated the patterns to capture driver and executor emitted metrics.

Compatible with Spark 2.0 / Spark 2.1 

No longer compatible Spark 1.6 and older.

### CODE REVIEW

File: spectator-ext-spark/src/main/resources/reference.conf [hunk 1]
```diff
@@ -2,14 +2,13 @@
 spectator.spark {
   name-patterns = [
     // EXECUTORS
-    // 97278898-4bd4-49c2-9889-aa5f969a7816-0013.97278898-4bd4-49c2-9889-aa5f969a7816-S1/2.executor.filesystem.file.largeRead_ops
+    // 97278898-4bd4-49c2-9889-aa5f969a7816-0013.2.executor.filesystem.file.largeRead_ops
```
Comment (zimmermatt) [inline spectator-ext-spark/src/main/resources/reference.conf]: To maintain compatibility, please add a block to support the new format and make the agent ID clause non-optional in the existing one.

For example (adjust as needed for valid `name` values in spark 2.+):
```
    // EXECUTORS
    // 97278898-4bd4-49c2-9889-aa5f969a7816-0013.97278898-4bd4-49c2-9889-aa5f969a7816-S1/2.executor.filesystem.file.largeRead_ops
    {
      pattern = "^([^.]+)\\.(([^.]+)\/)(\\d+)\\.((executor|jvm|CodeGenerator)\\..*)"
      name = 5
      tags = {
        "appId" = 1
        "agentId" = 3
        "executorId" = 4
      }
    },

    // 97278898-4bd4-49c2-9889-aa5f969a7816-0013.2.executor.filesystem.file.largeRead_ops
    {
      pattern = "^([^.]+)\\.(\\d+)\\.((executor|jvm|HiveExternalCatalog|CodeGenerator)\\..*)"
      name = 3
      tags = {
        "appId" = 1
        "executorId" = 2
      }
    },
```

File: spectator-ext-spark/src/test/java/com/netflix/spectator/spark/SparkNameFunctionTest.java [hunk 2]
```diff
@@ -41,32 +41,29 @@ private void assertEquals(Id expected, Id actual) {
   //EXECUTOR
   @Test
   public void executorMetric() {
```
Comment (zimmermatt) [inline spectator-ext-spark/src/test/java/com/netflix/spectator/spark/SparkNameFunctionTest.java]: Please keep the existing tests and add additional ones for all valid Spark 2.+ `name` values.

**Comment by brharrington** [comment]
Thanks for the updates.

Note, in the future, please fork the project and create PRs from the branch in your fork rather than pushing branches to the main project:

https://github.com/Netflix/spectator/blob/master/CONTRIBUTING.md#contributing-code

**Comment by pablete** [comment]
Ok. Thanks for the feedback. I will work on it in the next couple of days.

File: spectator-ext-spark/src/main/resources/reference.conf (hunk 1 above)
Comment (pablete) [reply spectator-ext-spark/src/main/resources/reference.conf]: Done

File: spectator-ext-spark/src/test/java/com/netflix/spectator/spark/SparkNameFunctionTest.java (hunk 2 above)
Comment (pablete) [reply spectator-ext-spark/src/test/java/com/netflix/spectator/spark/SparkNameFunctionTest.java]: Done

File: spectator-ext-spark/src/main/resources/reference.conf [hunk 3]
```diff
@@ -1,26 +1,38 @@
...
-        "executorId" = 4
+        "agentId" = 2
+        "executorId" = 3
       }
     },
 
+    // EXECUTORS Spark 2.0/2.1
+    // 97278898-4bd4-49c2-9889-aa5f969a7816-0013.2.executor.filesystem.file.largeRead_ops
+    {
+      pattern = "^([^.]+)\\.(\\d+)\\.((executor|jvm|HiveExternalCatalog|CodeGenerator)\\..*)"
```
Comment (zimmermatt) [inline spectator-ext-spark/src/main/resources/reference.conf]: On second look, the regex seems inconsistent with the example (here and the original). From the sample values it looks as though `executor` is first and required, followed by one of the following values. However, the regex says one of `executor`, `jvm`, `HiveExternalCatalog `, or `CodeGenerator`, followed by `.` and then anything.

Seems we may be missing some negative test cases.

Please let me know if there's something I'm missing.  If not, please correct and add negative tests for the 2.+ cases.

If you want to fix and add negative tests for 1.6, happy to accept the contribution :-), but not requesting that of you (one of us on telemetry can fill those in).

Comment (pablete) [reply spectator-ext-spark/src/main/resources/reference.conf]: I do not understand, do you want me to enumerate all possible examples for metrics emitted by the executors?

File: spectator-ext-spark/src/test/java/com/netflix/spectator/spark/SparkNameFunctionTest.java [hunk 4]
```diff
@@ -72,6 +92,16 @@ public void executorCodeGenerator() {
...
+    final Id expected = registry.createId("spark.CodeGenerator.compilationTime")
+        .withTag("appId", "97278898-4bd4-49c2-9889-aa5f969a7816-0013")
+        .withTag("executorId", "2")
+        .withTag("role", "executor");
+    assertEquals(expected, f.apply(name));
+  }
+
   // DRIVER
   @Test
   public void driverMetric() {
```
[the remark this answers is not in the document]
Comment (pablete) [reply spectator-ext-spark/src/test/java/com/netflix/spectator/spark/SparkNameFunctionTest.java]: same here, should I include all strings that can be captured by the regex patterns?

Ok. a bit overkill but I'll do it.

**Comment by pablete** [comment]
Updated with changes requested.

File: spectator-ext-spark/src/test/java/com/netflix/spectator/spark/SparkNameFunctionTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pablete) [reply spectator-ext-spark/src/test/java/com/netflix/spectator/spark/SparkNameFunctionTest.java]: Only provided one of each.

File: spectator-ext-spark/src/main/resources/reference.conf (hunk 3 above)
Comment (zimmermatt) [reply spectator-ext-spark/src/main/resources/reference.conf]: Sorry it wasn't clear. The additional example IDs you provided help.

I'm not familiar with the format of these IDs and my google search for a format spec didn't turn anything up (is there one available somewhere?). From the original example ids, it seemed like the pattern should have been specified as:
,,,
      pattern = "^([^.]+)\\.(\\d+)\\.(executor)\\.((jvm|HiveExternalCatalog|CodeGenerator)\\..*)"
,,,

However, the additional example IDs you provided clear that up and show that it is correct as is. So, thanks for adding those.

**Comment by zimmermatt** [comment]
Thanks again for the contribution, Pablo.

**Pull request merged**